### ISSUE TITLE
FirebaseMessaging is missing 'GetInstance' with FirebaseApp as parameter

### ISSUE THREAD

**Issue by haagOS** [xamarin/GooglePlayServicesComponents#438]
### Describe your Issue

The FirebaseMessaging binding library does not provide a way to retrieve the instance with a FirebaseApp as parameter.
According to the Android docs this should be possible.

See: https://firebase.google.com/docs/reference/admin/java/reference/com/google/firebase/messaging/FirebaseMessaging

> You can get an instance of FirebaseMessaging via getInstance(FirebaseApp), and then use it to send messages or manage FCM topic subscriptions.

> getInstance(FirebaseApp app)
Gets the FirebaseMessaging instance for the specified FirebaseApp.`

`FirebaseMessaging` (Xamarin.Firebase.Messaging) instead only has a static `Instance` property.
In contrast, for example, have a look at `FirebaseInstallations` - there is a static `Instance` property as well as a `GetInstance(FirebaseApp app)` method.

**My question: Is there something missing in the binding? And is there any way I can retrieve the FirebaseMessaging instance for a particular FirebaseApp?**

### Relevant information

Packages used:
```
    <PackageReference Include="Xamarin.Firebase.Messaging">
      <Version>121.0.1</Version>
    </PackageReference>
    <PackageReference Include="Xamarin.Forms">
      <Version>5.0.0.2012</Version>
    </PackageReference>
    <PackageReference Include="Xamarin.GooglePlayServices.Base">
      <Version>117.6.0</Version>
    </PackageReference>
```

### Minimal reproduction
```
var instance = Firebase.Messaging.FirebaseMessaging.Instance; // OK

var app = Firebase.FirebaseApp.GetInstance("some_id");
var specificInstance = Firebase.Messaging.FirebaseMessaging.GetInstance(app); // error: FirebaseMessaging does not contain a definition for 'GetInstance'

var installations = Firebase.Installations.FirebaseInstallations.GetInstance(app); // OK
```

**Comment by moljac** [comment]
Automatic tooling generates .NET properties for getter and/or setter. So, GetX ans SetX would end up as property X.

The problem is that there are more than 1 getter and they have different parameter types (overloads), so solution is to suppress property generation.
   
https://docs.microsoft.com/en-us/xamarin/android/platform/binding-java-library/customizing-bindings/java-bindings-metadata#propertyname

```
    <class abstract="false" deprecated="not deprecated" extends="java.lang.Object" extends-generic-aware="java.lang.Object" jni-extends="Ljava/lang/Object;" final="false" name="FirebaseMessaging" static="false" visibility="public" jni-signature="Lcom/google/firebase/messaging/FirebaseMessaging;">
      ....
      <method abstract="false" deprecated="not deprecated" final="false" name="getInstance" jni-signature="()Lcom/google/firebase/messaging/FirebaseMessaging;" bridge="false" native="false" return="com.google.firebase.messaging.FirebaseMessaging" jni-return="Lcom/google/firebase/messaging/FirebaseMessaging;" return-not-null="true" static="true" synchronized="true" synthetic="false" visibility="public">
      </method>
     ....
    </class>
```

In `api.xml` there is only one `getInstance` with no parameters. 

Decompiled `classes.jar`
 
```
package com.google.firebase.messaging;

import java.util.concurrent.*;
import com.google.android.datatransport.*;
import android.annotation.*;
import com.google.android.gms.common.internal.*;
import androidx.annotation.*;
import com.google.firebase.inject.*;
import com.google.firebase.platforminfo.*;
import com.google.firebase.heartbeatinfo.*;
import com.google.firebase.installations.*;
import com.google.android.gms.tasks.*;
import android.text.*;
import android.app.*;
import android.os.*;
import com.google.firebase.iid.*;
import javax.annotation.concurrent.*;
import com.google.firebase.*;
import android.content.*;
import android.content.pm.*;
import com.google.firebase.events.*;

public class FirebaseMessaging
{
    @Deprecated
    public static final String INSTANCE_ID_SCOPE = "FCM";
    private final Context context;
    private final FirebaseApp firebaseApp;
    private final FirebaseInstanceId iid;
    private final AutoInit autoInit;
    private final Executor fileIoExecutor;
    private final Task<TopicsSubscriber> topicsSubscriberTask;
    @Nullable
    @SuppressLint({ "FirebaseUnknownNullness" })
    @VisibleForTesting
    static TransportFactory transportFactory;
    
    @NonNull
    public static synchronized FirebaseMessaging getInstance() {
        return getInstance(FirebaseApp.getInstance());
    }
    
    @Keep
    @NonNull
    static synchronized FirebaseMessaging getInstance(@NonNull final FirebaseApp firebaseApp) {
        final FirebaseMessaging firebaseMessaging = (FirebaseMessaging)firebaseApp.get((Class)FirebaseMessaging.class);
        Preconditions.checkNotNull((Object)firebaseMessaging, (Object)"Firebase Messaging component is not present");
        return firebaseMessaging;
    }
    
    FirebaseMessaging(final FirebaseApp firebaseApp, final FirebaseInstanceId iid, final Provider<UserAgentPublisher> provider, final Provider<HeartBeatInfo> provider2, final FirebaseInstallationsApi firebaseInstallationsApi, @Nullable final TransportFactory transportFactory, final Subscriber subscriber) {
        try {
            Class.forName("com.google.firebase.iid.FirebaseInstanceIdReceiver");
        }
        catch (ClassNotFoundException ex) {
            throw new IllegalStateException("FirebaseMessaging and FirebaseInstanceId versions not compatible. Update to latest version of firebase-messaging.");
        }
        FirebaseMessaging.transportFactory = transportFactory;
        this.firebaseApp = firebaseApp;
        this.iid = iid;
        this.autoInit = new AutoInit(subscriber);
        this.context = firebaseApp.getApplicationContext();
        (this.fileIoExecutor = FcmExecutors.newInitExecutor()).execute(new FirebaseMessaging$$Lambda$0(this, iid));
        (this.topicsSubscriberTask = TopicsSubscriber.createInstance(firebaseApp, iid, new Metadata(this.context), provider, provider2, firebaseInstallationsApi, this.context, FcmExecutors.newTopicsSyncExecutor())).addOnSuccessListener(FcmExecutors.newTopicsSyncTriggerExecutor(), (OnSuccessListener)new FirebaseMessaging$$Lambda$1(this));
    }
    
    public boolean isAutoInitEnabled() {
        return this.autoInit.isEnabled();
    }
    
    public void setAutoInitEnabled(final boolean enabled) {
        this.autoInit.setEnabled(enabled);
    }
    
    @NonNull
    public boolean deliveryMetricsExportToBigQueryEnabled() {
        return MessagingAnalytics.deliveryMetricsExportToBigQueryEnabled();
    }
    
    public void setDeliveryMetricsExportToBigQuery(final boolean deliveryMetricsExportToBigQuery) {
        MessagingAnalytics.setDeliveryMetricsExportToBigQuery(deliveryMetricsExportToBigQuery);
    }
    
    @NonNull
    public Task<String> getToken() {
        return (Task<String>)this.iid.getInstanceId().continueWith(FirebaseMessaging$$Lambda$2.$instance);
    }
    
    @NonNull
    public Task<Void> deleteToken() {
        final TaskCompletionSource taskCompletionSource = new TaskCompletionSource();
        FcmExecutors.newNetworkIOExecutor().execute(new FirebaseMessaging$$Lambda$3(this, taskCompletionSource));
        return (Task<Void>)taskCompletionSource.getTask();
    }
    
    @NonNull
    public Task<Void> subscribeToTopic(@NonNull final String s) {
        return (Task<Void>)this.topicsSubscriberTask.onSuccessTask((SuccessContinuation)new FirebaseMessaging$$Lambda$4(s));
    }
    
    @NonNull
    public Task<Void> unsubscribeFromTopic(@NonNull final String s) {
        return (Task<Void>)this.topicsSubscriberTask.onSuccessTask((SuccessContinuation)new FirebaseMessaging$$Lambda$5(s));
    }
    
    public void send(@NonNull final RemoteMessage remoteMessage) {
        if (TextUtils.isEmpty((CharSequence)remoteMessage.getTo())) {
            throw new IllegalArgumentException("Missing 'to'");
        }
        final Intent intent = new Intent("com.google.android.gcm.intent.SEND");
        final Intent intent2 = new Intent();
        intent2.setPackage("com.google.example.invalidpackage");
        intent.putExtra("app", (Parcelable)PendingIntent.getBroadcast(this.context, 0, intent2, 0));
        intent.setPackage("com.google.android.gms");
        remoteMessage.populateSendMessageIntent(intent);
        this.context.sendOrderedBroadcast(intent, "com.google.android.gtalkservice.permission.GTALK_SERVICE");
    }
    
    @Nullable
    public static TransportFactory getTransportFactory() {
        return FirebaseMessaging.transportFactory;
    }
    
    private class AutoInit
    {
        private final Subscriber subscriber;
        @GuardedBy("this")
        private boolean initialized;
        @GuardedBy("this")
        @Nullable
        private EventHandler<DataCollectionDefaultChange> dataCollectionDefaultChangeEventHandler;
        @GuardedBy("this")
        @Nullable
        private Boolean autoInitEnabled;
        
        AutoInit(final Subscriber subscriber) {
            this.subscriber = subscriber;
        }
        
        synchronized void initialize() {
            if (this.initialized) {
                return;
            }
            this.autoInitEnabled = this.readEnabled();
            if (this.autoInitEnabled == null) {
                this.dataCollectionDefaultChangeEventHandler = (EventHandler<DataCollectionDefaultChange>)new FirebaseMessaging$AutoInit$$Lambda$0(this);
                this.subscriber.subscribe((Class)DataCollectionDefaultChange.class, (EventHandler)this.dataCollectionDefaultChangeEventHandler);
            }
            this.initialized = true;
        }
        
        synchronized boolean isEnabled() {
            this.initialize();
            final Boolean autoInitEnabled = this.autoInitEnabled;
            boolean b;
            if (autoInitEnabled != null) {
                b = autoInitEnabled;
            }
            else {
                b = FirebaseMessaging.this.firebaseApp.isDataCollectionDefaultEnabled();
            }
            return b;
        }
        
        synchronized void setEnabled(final boolean b) {
            this.initialize();
            final EventHandler<DataCollectionDefaultChange> dataCollectionDefaultChangeEventHandler = this.dataCollectionDefaultChangeEventHandler;
            if (dataCollectionDefaultChangeEventHandler != null) {
                this.subscriber.unsubscribe((Class)DataCollectionDefaultChange.class, (EventHandler)dataCollectionDefaultChangeEventHandler);
                this.dataCollectionDefaultChangeEventHandler = null;
            }
            final SharedPreferences.Editor edit = FirebaseMessaging.this.firebaseApp.getApplicationContext().getSharedPreferences("com.google.firebase.messaging", 0).edit();
            edit.putBoolean("auto_init", b);
            edit.apply();
            if (b) {
                FirebaseMessaging.this.fileIoExecutor.execute(new FirebaseMessaging$AutoInit$$Lambda$1(this));
            }
            this.autoInitEnabled = b;
        }
        
        @Nullable
        private Boolean readEnabled() {
            final Context applicationContext = FirebaseMessaging.this.firebaseApp.getApplicationContext();
            final SharedPreferences sharedPreferences = applicationContext.getSharedPreferences("com.google.firebase.messaging", 0);
            if (sharedPreferences.contains("auto_init")) {
                return sharedPreferences.getBoolean("auto_init", false);
            }
            try {
                final PackageManager packageManager = applicationContext.getPackageManager();
                if (packageManager != null) {
                    final ApplicationInfo applicationInfo = packageManager.getApplicationInfo(applicationContext.getPackageName(), 128);
                    if (applicationInfo != null && applicationInfo.metaData != null && applicationInfo.metaData.containsKey("firebase_messaging_auto_init_enabled")) {
                        return applicationInfo.metaData.getBoolean("firebase_messaging_auto_init_enabled");
                    }
                }
                return null;
            }
            catch (PackageManager.NameNotFoundException ex) {
                return null;
            }
        }
    }
}
``` 

In `api.xml.class-parse` there is no `getInstance` with `@NonNull final FirebaseApp firebaseApp` parameter.

Snippet from `api.xml.class-parse` for class `FirebaseMessaging`:

```
    <class
      abstract="false"
      deprecated="not deprecated"
      jni-extends="Ljava/lang/Object;"
      extends="java.lang.Object"
      extends-generic-aware="java.lang.Object"
      final="false"
      name="FirebaseMessaging"
      jni-signature="Lcom/google/firebase/messaging/FirebaseMessaging;"
      source-file-name="com.google.firebase:firebase-messaging@@21.0.1"
      static="false"
      visibility="public">
      <method
        abstract="false"
        deprecated="not deprecated"
        final="false"
        name="deleteToken"
        native="false"
        return="com.google.android.gms.tasks.Task&lt;java.lang.Void&gt;"
        jni-return="Lcom/google/android/gms/tasks/Task&lt;Ljava/lang/Void;&gt;;"
        static="false"
        synchronized="false"
        visibility="public"
        bridge="false"
        synthetic="false"
        jni-signature="()Lcom/google/android/gms/tasks/Task;"
        return-not-null="true" />
      <method
        abstract="false"
        deprecated="not deprecated"
        final="false"
        name="deliveryMetricsExportToBigQueryEnabled"
        native="false"
        return="boolean"
        jni-return="Z"
        static="false"
        synchronized="false"
        visibility="public"
        bridge="false"
        synthetic="false"
        jni-signature="()Z"
        return-not-null="true" />
      <method
        abstract="false"
        deprecated="not deprecated"
        final="false"
        name="getInstance"
        native="false"
        return="com.google.firebase.messaging.FirebaseMessaging"
        jni-return="Lcom/google/firebase/messaging/FirebaseMessaging;"
        static="true"
        synchronized="true"
        visibility="public"
        bridge="false"
        synthetic="false"
        jni-signature="()Lcom/google/firebase/messaging/FirebaseMessaging;"
        return-not-null="true" />
      <method
        abstract="false"
        deprecated="not deprecated"
        final="false"
        name="getToken"
        native="false"
        return="com.google.android.gms.tasks.Task&lt;java.lang.String&gt;"
        jni-return="Lcom/google/android/gms/tasks/Task&lt;Ljava/lang/String;&gt;;"
        static="false"
        synchronized="false"
        visibility="public"
        bridge="false"
        synthetic="false"
        jni-signature="()Lcom/google/android/gms/tasks/Task;"
        return-not-null="true" />
      <method
        abstract="false"
        deprecated="not deprecated"
        final="false"
        name="getTransportFactory"
        native="false"
        return="com.google.android.datatransport.TransportFactory"
        jni-return="Lcom/google/android/datatransport/TransportFactory;"
        static="true"
        synchronized="false"
        visibility="public"
        bridge="false"
        synthetic="false"
        jni-signature="()Lcom/google/android/datatransport/TransportFactory;" />
      <method
        abstract="false"
        deprecated="not deprecated"
        final="false"
        name="isAutoInitEnabled"
        native="false"
        return="boolean"
        jni-return="Z"
        static="false"
        synchronized="false"
        visibility="public"
        bridge="false"
        synthetic="false"
        jni-signature="()Z" />
      <method
        abstract="false"
        deprecated="not deprecated"
        final="false"
        name="send"
        native="false"
        return="void"
        jni-return="V"
        static="false"
        synchronized="false"
        visibility="public"
        bridge="false"
        synthetic="false"
        jni-signature="(Lcom/google/firebase/messaging/RemoteMessage;)V">
        <parameter
          name="message"
          type="com.google.firebase.messaging.RemoteMessage"
          jni-type="Lcom/google/firebase/messaging/RemoteMessage;"
          not-null="true" />
      </method>
      <method
        abstract="false"
        deprecated="not deprecated"
        final="false"
        name="setAutoInitEnabled"
        native="false"
        return="void"
        jni-return="V"
        static="false"
        synchronized="false"
        visibility="public"
        bridge="false"
        synthetic="false"
        jni-signature="(Z)V">
        <parameter
          name="enable"
          type="boolean"
          jni-type="Z" />
      </method>
      <method
        abstract="false"
        deprecated="not deprecated"
        final="false"
        name="setDeliveryMetricsExportToBigQuery"
        native="false"
        return="void"
        jni-return="V"
        static="false"
        synchronized="false"
        visibility="public"
        bridge="false"
        synthetic="false"
        jni-signature="(Z)V">
        <parameter
          name="p0"
          type="boolean"
          jni-type="Z" />
      </method>
      <method
        abstract="false"
        deprecated="not deprecated"
        final="false"
        name="subscribeToTopic"
        native="false"
        return="com.google.android.gms.tasks.Task&lt;java.lang.Void&gt;"
        jni-return="Lcom/google/android/gms/tasks/Task&lt;Ljava/lang/Void;&gt;;"
        static="false"
        synchronized="false"
        visibility="public"
        bridge="false"
        synthetic="false"
        jni-signature="(Ljava/lang/String;)Lcom/google/android/gms/tasks/Task;"
        return-not-null="true">
        <parameter
          name="topic"
          type="java.lang.String"
          jni-type="Ljava/lang/String;"
          not-null="true" />
      </method>
      <method
        abstract="false"
        deprecated="not deprecated"
        final="false"
        name="unsubscribeFromTopic"
        native="false"
        return="com.google.android.gms.tasks.Task&lt;java.lang.Void&gt;"
        jni-return="Lcom/google/android/gms/tasks/Task&lt;Ljava/lang/Void;&gt;;"
        static="false"
        synchronized="false"
        visibility="public"
        bridge="false"
        synthetic="false"
        jni-signature="(Ljava/lang/String;)Lcom/google/android/gms/tasks/Task;"
        return-not-null="true">
        <parameter
          name="topic"
          type="java.lang.String"
          jni-type="Ljava/lang/String;"
          not-null="true" />
      </method>
      <field
        deprecated="deprecated"
        final="true"
        name="INSTANCE_ID_SCOPE"
        static="true"
        synthetic="false"
        transient="false"
        type="java.lang.String"
        type-generic-aware="java.lang.String"
        jni-signature="Ljava/lang/String;"
        value="&quot;FCM&quot;"
        visibility="public"
        volatile="false" />
    </class>

```

**Comment by jpobst** [comment]
Per the decompiled code, this method is `private`, which `class-parse` will not emit:
```
@Keep
@NonNull
static synchronized FirebaseMessaging getInstance(@NonNull final FirebaseApp firebaseApp) {
    final FirebaseMessaging firebaseMessaging = (FirebaseMessaging)firebaseApp.get((Class)FirebaseMessaging.class);
    Preconditions.checkNotNull((Object)firebaseMessaging, (Object)"Firebase Messaging component is not present");
    return firebaseMessaging;
}
```
I guess the interesting question is why do the Firebase docs say it is `public` when the binary says it is `private`?

**Comment by moljac** [comment]
@haagOS I guess you'll have to ask google about this issue

**Comment by haagOS** [comment]
Thanks you very much for your effort.

In response to your answers I have searched through the [firebase-android-sdk](https://github.com/firebase/firebase-android-sdk) and found this issue [Allow FirebaseMessaging to be initialized with a secondary FirebaseApp](https://github.com/firebase/firebase-android-sdk/issues/2285#issue-775004480) that is quite identical to my issue described. 
Within that issue there is a link to stackoverflow containing further details including an answer from the firebase support team: [FirebaseMessaging.getInstance(firebaseApp) for secondary app supposed to be public but it's private?](https://stackoverflow.com/questions/62253691/firebasemessaging-getinstancefirebaseapp-for-secondary-app-supposed-to-be-publ)

So this is clearly not an issue with the binding. Closing this one.
